ra/ra.go

```diff
@@ -958,27 +969,59 @@ func (ra *RegistrationAuthorityImpl) FinalizeOrder(ctx context.Context, req *rap
 		}
 	}
 
+	err = processOrder(ctx, ra, order, csrOb)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note how many names were in this finalized certificate order.
+	ra.namesPerCert.With(
+		prometheus.Labels{"type": "issued"},
+	).Observe(float64(len(order.Names)))
+
+	// Update the order status locally since the SA doesn't return the updated
+	// order itself after setting the status
+	order.Status = string(core.StatusValid)
+	return order, nil
+}
+
+// Process the order through the StorageAuthority and issue a certificate.
+// If any error happens during the process, fail the order and return the error.
+func processOrder(
+	ctx context.Context,
+	ra *RegistrationAuthorityImpl,
+	order *corepb.Order,
+	csrOb *x509.CertificateRequest) error {
+	// Create a goroutine to handle marking the order as failed.
+	// We attach a separate context to prevent that the original context expiration
+	// prevents the order to be marked as failed.
+	problems := make(chan *probs.ProblemDetails)
+	failCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	go ra.failOrder(failCtx, order.Id, problems)
+	defer close(problems)
 	// Update the order to be status processing - we issue synchronously at the
 	// present time so this is somewhat artificial/unnecessary but allows planning
 	// for the future.
 	//
-	// NOTE(@cpu): After this point any errors that are encountered must update
-	// the state of the order to invalid by setting the order's error field.
+	// NOTE(@cpu): After this point any errors that are encountered must notify
+	// that the state of the order is invalid by sending the order's error in the
+	// problems channel.
 	// Otherwise the order will be "stuck" in processing state. It can not be
 	// finalized because it isn't pending, but we aren't going to process it
 	// further because we already did and encountered an error.
-	_, err = ra.SA.SetOrderProcessing(ctx, &sapb.OrderRequest{Id: order.Id})
+	_, err := ra.SA.SetOrderProcessing(ctx, &sapb.OrderRequest{Id: order.Id})
 	if err != nil {
 		// Fail the order with a server internal error - we weren't able to set the
 		// status to processing and that's unexpected & weird.
-		ra.failOrder(ctx, order, probs.ServerInternal("Error setting order processing"))
-		return nil, err
+		problems <- probs.ServerInternal("Error setting order processing")
+		return err
 	}
 
 	// Attempt issuance for the order. If the order isn't fully authorized this
 	// will return an error.
 	issueReq := core.CertificateRequest{
-		Bytes: req.Csr,
+		Bytes: csrOb.Raw,
```

Comment:
Can you please confirm this is exact?
All tests passes, I figured out it was not necessary to pass both the request and the CSR built off it, but I am not entirely sure

Reply:
Yes, `req.Csr` and `csrOb.Raw` should have the exact same bytes. But there's no reason to change this in this PR (it's unrelated to the main change you're making) so I'd ask that you don't change it now. If you'd like to open up a follow-up PR that changes this, you're welcome to do so!